src/syft/lib/misc/union.py

```diff
@@ -52,12 +52,12 @@ def from_qualnames(union_types: List[str]) -> str:
             return name
 
         target_type = UnionGenerator(name, tuple(), {})
-        union_cache.add(name)
 
         # adding the type and it's union types to a lazy dict that will be solved
         # when the ast is created.
         lazy_pairing[target_type] = union_types
         qualname = target_type.__qualname__
+        union_cache.add(qualname)
```

Comment:
Hmmm, you may by right, but why? Is there a case where there are created same unions with different paths?

Reply:
It was broken because it cached the name not the qual name and the resultant name couldn't resolve the full Union class path, but it didn't happen until we created two of the same Union which my PR required in the allowlist.

```
-        union_cache.add(name)
+        union_cache.add(qualname)
```

Reply:
OOH, cool!